core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/ProgressEventListener.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.progress;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.progress.AbstractProgressEventListenerBuilder.DefaultProgressEventListenerBuilder;
+import software.amazon.awssdk.core.progress.ProgressEventHandler.ByteCountEventHandler;
+import software.amazon.awssdk.core.progress.ProgressEventHandler.RequestLifeCycleEventHandler;
+
+/**
+ * Listener interface to listen to the progress events.
+ *
+ * <T> Type of progress event
+ *
+ * @see ProgressEvent the progress event
+ */
+@SdkPublicApi
+public interface ProgressEventListener extends ProgressEventHandler, ByteCountEventHandler, RequestLifeCycleEventHandler {
+
+    Set<Class<? extends ProgressEvent>> SUPPORTED_TYPES =
+        Collections.unmodifiableSet(new HashSet<>(Arrays.asList(RequestCycleEvent.class, ByteCountEvent.class)));
+
+    /**
+     * Check if the provided event is supported by this listener or not.
+     *
+     * <p>
+     * Sub classes can override {@link #extendedTypes()} to provided additional supported {@link ProgressEvent}.
+     *
+     * @param event the event
+     * @return true if the event is supported by this listener, otherwise false.
+     */
+    default boolean isSupported(ProgressEvent event) {
+        return SUPPORTED_TYPES.contains(event.getClass()) || extendedTypes().contains(event.getClass());
+    }
+
+    /**
+     * Data notification when a {@link ProgressEvent} is available.
+     *
+     * @param progressEvent a progress event
+     * @return the future containing the {@link ProgressEventResult}
+     */
+    default CompletableFuture<? extends ProgressEventResult> onProgressEvent(ProgressEvent progressEvent) {
+
+        ProgressEventType progressEventType = progressEvent.eventType();
+
+        if (progressEventType instanceof ByteCountEventType) {
```

Comment:
Yes, visitor pattern requires the visitor to know all sub-types of the events and would not work for extended events.

Taking `S3TransferEvent` for example, 

```
    @Override
    public void accept(ProgressEventHandler visitor) {
        visitor.visit(this); // would not compile because the core`ProgressEventHandler` does not know about `S3TransferEvent`
    }
```